inputs.md
======

| Flags | Required? | Description |
|-|-|-|
| -h <file> | Required | "Haplotypes"<br><br>Binary representation (as allele) of haplotypes.<br>  **Rows** - SNPs<br>  **Cols** - Haplotype  |
    | -l <file> | Required | "Legend"<br><br>Legend file for SNP markers<br>Rows - SNPs << 1:1 with haplotype file<br>Cols -<br>  1. rsID<br>  2. base pair position<br>  3. base allele rep. by 0<br>  4. base allele rep. by 1<br><br>first line are column labels/header |
| -m <file> | Required | "Map"<br><br>fine-scale recombination rate across the region.<br>Rows - SNPs << 1:1 with haplotype file?<br>Cols - <br>  1. position <br><br>  2. COMBINED_rate(cM/Mb) (right of marker)<br><br>  3. Genetic_Map(cM) (cumulative, left of marker) |
| -dl <int> <a> <rr1> <rr2> | Required | "Disease Loci"<br><br>location, risk allele, relative risk for each disease risk.<br>  1. physical location of SNP, which must be in legend file supplied to the -l flag                <br>  2. risk allele (0 or 1), the corresponding base can found in the legend <br>  3. heterozygote disease risk <br>  4. homozygote disease risk<br><br>For example, ```-dl 1085679 1 1.5 2.25 2190692 0 2 4``` specifies two disease SNPs, at positions <br><br>1085679 and 2190692, and with heterozgyote risks 1.5 and 2, homozygote risks 2.25 and 4, and risk <br><br>alleles set to 1 and 0 at each SNP respectively. There is no limit on the number of disease SNPs. <br><br>We simulate under a disease model where the disease SNPs are independent, and the haplotypes defined <br><br>by the disease SNPs are in HWE. <br><br><br>This flag is optional for version 2.0.2 and above, when if not supplied then all haplotypes will be simulated under the null. |
| -n <int> <int> | Recommended | Sets the number of control, and the number of case individuals, to simulate. <br><br>e.g. ```-n 100 200``` simulates 100 control and 200 case individuals.  Default is to generate 1 control and 1 case individual.  |
| -int <int> <int>  | Optional | Specify the lower and upper boundaries of the region in which you wish to carry out simulation.  The default is set to 0...500000000 |
| -o <file>  | Required | Output file prefix.  For example ```-o ex.out[.gz]``` creates the following files for the case data:<br><br>* ```ex.out.cases.haps[.gz]``` - a file containing the simulated haplotype data in the same format as the haplotype file provided by `-h` flag.<br><br>* ```ex.out.legend``` (v2.1.2+) - A legend file with information about the SNPs in the .haps files. <br><br><br>* ```ex.out.cases.gen[.gz]``` - A file containing the simulated genotype data, in file format compatible with {SNPTEST, SNPTEST2, IMPUTE, IMPUTE2, GTOOL}<br><br>* ```ex.out.cases.sample``` - A sample file in the file format compatible with SNPTEST2 for the simulated genotype data. <br><br><br>* ```ex.out.cases.tags.gen[.gz]``` - Genotype data limited to the subset of SNPs specified by the file supplied to the `-t` flag (if applicable). <br><br><br>A similar set of files will be produced for the control data, with the same file names, except that `cases` are replaced by `controls`.<br>A summary file, ```ex.out.[.gz]summary```, will also by produced, which summarizes the simulation parameters, input files, and output files. <br><br><br>Note:<br>* If the output file prefix has a `.gz` extension, then the *.haps, *.gen, *.tags.gen will be gzipped. <br><br>* It is possible to suppress some of the output files using the flags `-no_gens_output` and `-no_haps_output`<br>`-h` flag  |
| -output_snp_summary | Optional | Output the p-values and effect-size estimates (under a log additive model test) for each disease SNP and under a joint model for all of the disease SNPs in the simulated genotype data.   |
| -no_haps_output | Optional | No haplotype data files, `*.haps[.gz]`, will be output for the case and control data.  |
| -no_gens_output | Optional | No genotype files, `*.haps[.gz]`, will be output for the case and control data.<br><br>However, if you have provided an input to the `-t` flag, then the `*.tags.gen[.gz]` will be output. |
| -t <file>  | Optional | "Tags"<br><br>Specify subset of SNP markers to output from simulation - position, one per line. <br>- Rows: base pair position/physical location<br><br>Notes:<br>* The physical locations of the given SNP markers must match those in the legends file. <br>* If this option is selected, then a `.tags.gen` output will be produced, that contains the positions of the SNPs in the output file.  |
| -Ne <int> | Optional | Sets effective population size that scales the fine-scale recombination map for the given population. <br><br>e.g. ```-Ne 11000``` sets the effective population size to 11000. <br><br>Notes:<br>* For autosomal chromosomes, highly recommend the values<br>    - 11418 for CEPH<br>    - 17469 for Yoruban<br>    - 14269 for Chinese & Japanese populations |
| -theta <real>  | Optional | Mutation rate in the model. <br><br>e.g. ```-theta 10``` sets the scaled mutation rate to 10.  <br><br>Notes:<br>* Mutation rate is set such that the expected number of mutations at a given SNP is equal to 1 by default. |